### DISCUSSION TITLE
Why the switch from Prepl to nRepl ?

### DISCUSSION THREAD

**Discussion by didibus** [Olical/conjure#337] [Q&A] [answered]
I just finished reading your blog post: https://oli.me.uk/exploring-repl-tooling-with-prepl/ and then went to checkout Conjure, and was surprised that the documentation now mentions Conjure uses nRepl for Clojure, Clojurescript, Babashka and even shadow-cljs.

I feel that deserves a follow up blog posts haha

Anyways, I'm very curious, was there too many pain points in the end with Prepl? Would you advise against it now for other tools authors?

**Comment by Olical** [comment, marked as answer]
So the TL;DR last time I wrote all of this down was essentially: prepl is hyper minimal and good when you need a simple thing that lets you automate sending code from one program to another.

It falls down when you try to do anything interesting from a user experience and developer tool standpoint. I spent a couple of *years* getting it powerful enough to work with and it still massively lacked features that nREPL has as standard as well as didn't really work at all for ClojureScript. I would've had to spend more years rewriting it and essentially reimplementing nREPL (which at the time, I didn't like because I thought it was system that broke all the time when it updated (this wasn't actually true)).

So eventually I gave in, I tried supporting nREPL in my next rewrite and everything worked beautifully. All aspects of Clojure were supported, I could provide far more developer tools for users with far less effort. I'm even going to integrate a debugger over the next few weeks which would never have been possible with prepl (well, maybe it would, but it'd be REALLY hard).

prepl solves the problem of connecting a simple CLI REPL or program to a remote Clojure program so you can send some code at it. nREPL, more specifically nREPL+CIDER, solves actual day to day developer experience challenges. We get to lean on the collective effort of the entire Clojure tool developer community.

prepl was just me on my own trying my best to even match up one eval response to another and doing weird hacks to get some sort of autocompletion to work.

Since Conjure's design is so much better these days I can actually add support for prepl but as an opt in system, Conjure supports many languages and many transports. So there's nothing (other than time and effort) stopping me from implementing `conjure.client.clojure.prepl` alongside `conjure.client.clojure.nrepl`. It will however do barely anything in comparison to the nREPL version, maybe just simple evals.

I just recommend people use nREPL for anything a human needs to do. I'd recommend prepl when you need a machine to talk to a machine (which is very rare and only really useful in simple dev tool areas I think).

I hope this sort of answers it. I've answered this a few times and I remember writing all of this up somewhere but I can't find it, I guess this will have to serve as the answer now! Maybe it was a twitter thread from a few years back... if so I'll never find that.

**Reply by leafgarland** [reply]
I've spent a little time hacking together a conjure.client.clojure.prepl client. I'm not entirely sure why 🤨. Conjure and nrepl are working fine for me, but I'm tempted by the minimalism of using prepl and I've found that clojure-lsp does most of the extra features that nrepl provides (this is perhaps different to when you were creating the prepl version of conjure).

I'm also interested in some of the things raised in talks/articles like

- [REPL-Driven Development](https://github.com/matthiasn/talk-transcripts/blob/master/Halloway_Stuart/REPLDrivenDevelopment.md)
- [Rich Hickey talking about REPLs vs RPC systems](https://nextjournal.com/mk/rich-hickey-on-repls)
- [Customising your REPL in clj](https://insideclojure.org/2020/02/11/custom-repl/)

Something about minimal parts that we can combine in many ways to make things bigger than their parts. I'm still trying to understand it but I figured I might be able to work with prepl + clojure-lsp, and then see if I can find some enlightenment - or not 😄

**Reply by Olical** [reply]
Oh that's cool! I'd love to see it some day. Conjure's design allows us to include that as an optional client people switch on which wouldn't get in the way of the existing default nREPL implementation anyway. So if you get something working (even if it's really minimal and only allows basic evals) we could include it for others to try and improve over time.

I personally still prefer nREPL + CIDER for getting stuff done, but a hyper minimal prepl client can't hurt, the architecture is designed to support it alongside nREPL 😄

**Comment by Quantisan** [comment]
I came across this. I switched from conjure to using clojure-lsp exclusively but has been looking for a way to send code to a Clojure REPL. @leafgarland, did you get any further along a prepl + clojure-lsp setup?

**Comment by leafgarland** [comment]
@Quantisan I had a "working" prepl but with only a simple EDN parser. I was hoping to make it a streaming parser to cope with larger messages but didn't find the time.

**Reply by leafgarland** [reply]
I can't promise to move quickly on getting it out but if you are interested I will look at pushing what I have to a repo.
